### PR TITLE
Flatten metadata into a single list.

### LINKED ISSUE #34: Metadata files should be valid JSON documents
The current metadata file format is a sequence of lines, each of which is a valid JSON document that cannot contain the `\n` character.

Can we fix this so that the metadata file contains a valid JSON document? Probably just a JSON array of objects, each describing a feature.

I was stumped by this for a while trying to process coppersmith generated metadata files recently, and @laurencer has just run into it trying to use hand-written metadata (which, alas, had far, far, *far* too many new lines in it).

### PR DESCRIPTION
**Pull request by triggerNZ**
Fixes #34

### CODE REVIEW

**Comment by laurencer** [comment]
It might be useful having an example JSON output and including things like a `version` and maybe retaining the multiple feature sets but having them as named things - e.g.

```
{
  "version": "1.0",
  "feature-sets": [
    {
      "name": "my-features",
      "features": [ {...} ]
  ]
}
```

@thsutton - any comments?

**Comment by triggerNZ** [comment]
@laurencer thoughts on publishing and validating against a schema along the lines of http://json-schema.org/ ?

**Comment by laurencer** [comment]
@triggerNZ not sure that's needed though may be useful. I trust your judgement. It's probably enough to have a couple of Scala case classes which are the spec with argonaut serialisers (and keep multiple versions that you use for doing regression tests).

My only advice would be to focus on the customer. For people who are going to consume it, examples in the user guide, readme, or docs might be more useful than a JSON schema. At least internally, I think you can assume that the audience is mostly our internal A&I engineers who are familiar with at least one JSON library (Play JSON, Argonaut, etc) and and understand Scala data structures.

**Comment by laurencer** [comment]
@jameshales @mkiedys - can you comment on the metadata format/any improvements you'd like to see made?

**Comment by triggerNZ** [comment]
@laurencer I will raise a separate PR around a published JSON standard with validators and something we will commit to. This one was more around just producing valid JSON documents

File: plugin/src/main/scala/commbank/coppersmith/plugin/CoppersmithPlugin.scala [hunk 1]
```diff
@@ -55,10 +55,10 @@ object CoppersmithPlugin extends AutoPlugin {
         )!!
 
         if (res.length == 0) {
-          error("Metadata empty. Are your feature definitions in a subproject?")
+          sys.error("Metadata empty. Are your feature definitions in a subproject?")
         } else {
           val metadataFile = tgt / s"coppersmith-features-${v}.json"
-
+          println("Outputting " + res)
```
Comment (dkristian) [inline plugin/src/main/scala/commbank/coppersmith/plugin/CoppersmithPlugin.scala]: I think this should use SBT's logging, perhaps at `debug` level.

File: core/src/main/scala/commbank/coppersmith/MetadataOutput.scala [hunk 2]
```diff
@@ -78,10 +78,14 @@ object MetadataOutput {
     def metadata = mds.metadata
   }
 
+  def metadataObjects[S, O](metadata: List[(Metadata[S, Feature.Value], Option[Conforms[_, _]])], printer: MetadataPrinter[O]): List[O] = {
```
Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/MetadataOutput.scala]: Watch line length; could do with wrapping params as per `metadataString` below.

File: plugin/src/main/scala/commbank/coppersmith/plugin/CoppersmithPlugin.scala [hunk 3]
```diff
@@ -55,10 +55,10 @@ object CoppersmithPlugin extends AutoPlugin {
         )!!
 
         if (res.length == 0) {
-          error("Metadata empty. Are your feature definitions in a subproject?")
+          sys.error("Metadata empty. Are your feature definitions in a subproject?")
```
Comment (dkristian) [inline plugin/src/main/scala/commbank/coppersmith/plugin/CoppersmithPlugin.scala]: What's the effect of calling `sys.error` from a plugin? Is there a better error handling framework from SBT that could be used instead?

File: plugin/src/sbt-test/coppersmith-plugin/simple/src/main/scala/hello.scala [hunk 4]
```diff
@@ -38,3 +38,19 @@ object FluentMovieFeatures extends FeatureSetWithTime[Movie] {
...
+
+object FluentMovieFeatures2 extends FeatureSetWithTime[Movie] {
+  val namespace            = "plugin.examples"
+  def entity(movie: Movie) = movie.id
+
+  val source      = From[Movie]()  // FeatureSource (see above)
+  val select      = source.featureSetBuilder(namespace, entity)
+
+  val format      = DateTimeFormat.forPattern("dd-MMM-yyyy").withLocale(Locale.ENGLISH)
+
```
Comment (dkristian) [inline plugin/src/sbt-test/coppersmith-plugin/simple/src/main/scala/hello.scala]: Minor: vertical alignment

File: plugin/src/sbt-test/coppersmith-plugin/simple/src/main/scala/hello.scala [hunk 5]
```diff
@@ -38,3 +38,19 @@ object FluentMovieFeatures extends FeatureSetWithTime[Movie] {
...
+  val movieReleaseDay2  = select(_.releaseDate)
```
Comment (dkristian) [inline plugin/src/sbt-test/coppersmith-plugin/simple/src/main/scala/hello.scala]: Minor: remove additional space before `=`

File: plugin/src/sbt-test/coppersmith-plugin/simple/src/main/scala/hello.scala [hunk 6]
```diff
@@ -38,3 +38,19 @@ object FluentMovieFeatures extends FeatureSetWithTime[Movie] {
...
+    .asFeature(Nominal, "MOVIE_RELEASE_DAY2", "Day on which the movie was released - take 2")
+
+
```
Comment (dkristian) [inline plugin/src/sbt-test/coppersmith-plugin/simple/src/main/scala/hello.scala]: Minor: remove additional newline

File: tools/src/main/scala/commbank/coppersmith/tools/MetadataMain.scala [hunk 7]
```diff
@@ -19,22 +19,39 @@ import commbank.coppersmith.Feature.Conforms, Conforms.conforms_?
...
+  sealed trait FormatType
+  case object PsvFormat extends FormatType
+  case object JsonFormat extends FormatType
+
   def main(args:Array[String]) = {
     val (format, packagge) = args.take(2) match {
-      case Array("--psv", pkg)  => ( MetadataOutput.Psv, pkg)
-      case Array("--json", pkg) => ( MetadataOutput.JsonObject, pkg)
-      case Array(pkg)           => ( MetadataOutput.JsonObject, pkg)
+      case Array("--psv", pkg)  => ( PsvFormat, pkg)
```
Comment (dkristian) [inline tools/src/main/scala/commbank/coppersmith/tools/MetadataMain.scala]: Minor: remove additional space before `PsvFormat`. Can move to before `pkg` to align with below.

File: tools/src/main/scala/commbank/coppersmith/tools/MetadataMain.scala [hunk 8]
```diff
@@ -19,22 +19,39 @@ import commbank.coppersmith.Feature.Conforms, Conforms.conforms_?
...
+
     val allConforms =
       ObjectFinder.findObjects[Conforms[_, _]](args(0), "commbank.coppersmith", "au.com.cba.omnia")
 
-    metadataSets.foreach { ms =>
-      val metadataConformsSet = ms.metadata.map(m => (m, allConforms.find(c => conforms_?(c, m)))).toList
-      val outputString = MetadataOutput.metadataString(metadataConformsSet, format)
-      println(outputString)
+    // The repetition here is regrettable but getting the types without statically
+    //knowing the formatter is really awkward
```
Comment (dkristian) [inline tools/src/main/scala/commbank/coppersmith/tools/MetadataMain.scala]: Minor: space after `//`

**Comment by mkiedys** [comment]
@laurencer wrote:

> It might be useful having an example JSON output and including things like a version and maybe retaining the multiple feature sets but having them as named things - e.g.
👍 

I think people including myself would benefit if exact type would be provided including everything that is known at compile time about the type like: min length, max length, min value, max value, scale, precision, source timezone, null equivalents, is nullable, signed, unsigned etc.

**Comment by triggerNZ** [comment]
Closing this in favour of a new PR which addresses some of the issues raised by @laurencer

**Pull request closed without merging**